File: projects/coffee-bot/src/parser.py
import csv
import json
import re
from pathlib import Path

DRINK_KEYWORDS = {
    "latte": "Latte",
    "flat white": "Flat White",
    "americano": "Americano",
    "cappuccino": "Cappuccino",
    "matcha": "Matcha Latte",
    "mocha": "Mocha",
    "espresso": "Espresso",
}

MILK_KEYWORDS = {
    "whole": "Whole Milk",
    "full fat": "Whole Milk",
    "semi": "Semi-Skimmed Milk",
    "skim": "Skimmed Milk",
    "oat": "Oat Milk",
    "almond": "Almond Milk",
    "soy": "Soy Milk",
    "soya": "Soy Milk",
    "coconut": "Coconut Milk",
}

SIZE_KEYWORDS = {
    "small": "Small",
    "medium": "Medium",
    "large": "Large",
    "regular": "Medium",
}

DEFAULT_SIZE = "Large"
DEFAULT_MILK = "Whole Milk"


def normalize_text(text: str) -> str:
    return re.sub(r"\s+", " ", (text or "").strip().lower())


def detect_drink(text: str):
    hits = [v for k, v in DRINK_KEYWORDS.items() if k in text]
    if not hits:
        return None, 0.0
    # If multiple drinks hit, ambiguous
    if len(set(hits)) > 1:
        return "/".join(sorted(set(hits))), 0.35
    return hits[0], 0.85


def detect_milk(text: str):
    for k, v in MILK_KEYWORDS.items():
        if k in text:
            return v, False
    return DEFAULT_MILK, True


def detect_size(text: str):
    for k, v in SIZE_KEYWORDS.items():
        if k in text:
            return v, False
    return DEFAULT_SIZE, True


def parse_row(row):
    name = (row.get("name") or "").strip()
    raw = row.get("order_text") or ""
    qty = row.get("qty") or "1"
    text = normalize_text(raw)

    drink, drink_conf = detect_drink(text)
    size, size_defaulted = detect_size(text)
    milk, milk_defaulted = detect_milk(text)

    clarification_needed = False
    reason = ""

    if not text:
        clarification_needed = True
        reason = "blank_order"
        drink = "UNKNOWN"
        drink_conf = 0.0
    elif drink is None:
        clarification_needed = True
        reason = "no_drink_detected"
        drink = "UNKNOWN"
        drink_conf = 0.2
    elif "/" in drink:
        clarification_needed = True
        reason = "multiple_drink_candidates"

    confidence = round(min(0.99, drink_conf + (0.07 if not size_defaulted else 0) + (0.07 if not milk_defaulted else 0)), 2)

    return {
        "name": name,
        "raw_order_text": raw,
        "quantity": int(qty) if str(qty).isdigit() else 1,
        "drink": drink,
        "size": size,
        "milk": milk,
        "defaults_applied": {
            "size": size_defaulted,
            "milk": milk_defaulted,
        },
        "confidence": confidence,
        "clarification_needed": clarification_needed,
        "clarification_reason": reason,
    }


def run(input_csv: Path, output_json: Path):
    rows = []
    with input_csv.open("r", encoding="utf-8-sig", newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(parse_row(row))

    summary = {
        "total_orders": len(rows),
        "clarification_count": sum(1 for r in rows if r["clarification_needed"]),
        "auto_default_size_count": sum(1 for r in rows if r["defaults_applied"]["size"]),
        "auto_default_milk_count": sum(1 for r in rows if r["defaults_applied"]["milk"]),
    }

    payload = {"summary": summary, "orders": rows}
    output_json.parent.mkdir(parents=True, exist_ok=True)
    output_json.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"Parsed {summary['total_orders']} rows -> {output_json}")
    print(f"Clarifications needed: {summary['clarification_count']}")


if __name__ == "__main__":
    base = Path(__file__).resolve().parents[1]
    input_csv = base / "samples" / "blank_street_orders_sample.csv"
    output_json = base / "output" / "parsed_orders.json"
    run(input_csv, output_json)

File: projects/coffee-bot/src/parser.mjs
import fs from 'node:fs';
import path from 'node:path';

const MENU = JSON.parse(
  fs.readFileSync(path.resolve(process.cwd(), 'projects', 'coffee-bot', 'data', 'blank_street_menu.json'), 'utf8')
);

const MILK_KEYWORDS = new Map([
  ['whole', 'Whole Milk'],
  ['full fat', 'Whole Milk'],
  ['semi', 'Semi-Skimmed Milk'],
  ['skim', 'Skimmed Milk'],
  ['oat', 'Oat Milk'],
  ['almond', 'Almond Milk'],
  ['soy', 'Soy Milk'],
  ['soya', 'Soy Milk'],
  ['coconut', 'Coconut Milk'],
]);

const SIZE_KEYWORDS = new Map([
  ['small', 'Small'],
  ['medium', 'Medium'],
  ['large', 'Large'],
  ['regular', 'Medium'],
]);

const DEFAULT_SIZE = 'Large';
const DEFAULT_MILK = 'Whole Milk';

function normalizeText(text = '') {
  return text.trim().toLowerCase().replace(/\s+/g, ' ');
}

function parseCsv(content) {
  const lines = content.split(/\r?\n/).filter(Boolean);
  if (!lines.length) return [];
  const headers = lines[0].split(',').map(h => h.trim());
  const rows = [];

  for (let i = 1; i < lines.length; i += 1) {
    const line = lines[i];
    const fields = [];
    let current = '';
    let inQuotes = false;

    for (let j = 0; j < line.length; j += 1) {
      const ch = line[j];
      if (ch === '"') {
        inQuotes = !inQuotes;
      } else if (ch === ',' && !inQuotes) {
        fields.push(current);
        current = '';
      } else {
        current += ch;
      }
    }
    fields.push(current);

    const obj = {};
    headers.forEach((h, idx) => {
      obj[h] = (fields[idx] || '').trim();
    });
    rows.push(obj);
  }
  return rows;
}

function detectDrink(text) {
  const matches = [];
  for (const item of MENU.items) {
    for (const alias of item.aliases) {
      if (text.includes(alias)) {
        matches.push({ canonical: item.canonical, alias });
      }
    }
  }

  if (matches.length === 0) return { drink: null, conf: 0.0 };

  // Prefer the most specific alias (longest phrase), e.g. "pistachio latte" over "latte"
  matches.sort((a, b) => b.alias.length - a.alias.length);
  const top = matches[0];

  const distinctTopCanonicals = [...new Set(matches.filter(m => m.alias.length === top.alias.length).map(m => m.canonical))];
  if (distinctTopCanonicals.length > 1) {
    return { drink: distinctTopCanonicals.sort().join('/'), conf: 0.35 };
  }

  return { drink: top.canonical, conf: 0.92 };
}

function detectMilk(text) {
  for (const [k, v] of MILK_KEYWORDS.entries()) {
    if (text.includes(k)) return { milk: v, defaulted: false };
  }
  return { milk: DEFAULT_MILK, defaulted: true };
}

function detectSize(text) {
  for (const [k, v] of SIZE_KEYWORDS.entries()) {
    if (text.includes(k)) return { size: v, defaulted: false };
  }
  return { size: DEFAULT_SIZE, defaulted: true };
}

function parseRow(row) {
  const name = (row.name || '').trim();
  const raw = row.order_text || '';
  const qty = Number.parseInt(row.qty || '1', 10) || 1;
  const text = normalizeText(raw);

  const { drink: drinkDetected, conf: drinkConf } = detectDrink(text);
  const { size, defaulted: sizeDefaulted } = detectSize(text);
  const { milk, defaulted: milkDefaulted } = detectMilk(text);

  let drink = drinkDetected;
  let clarificationNeeded = false;
  let reason = '';

  if (!text) {
    clarificationNeeded = true;
    reason = 'blank_order';
    drink = 'UNKNOWN';
  } else if (!drinkDetected) {
    clarificationNeeded = true;
    reason = 'no_drink_detected';
    drink = 'UNKNOWN';
  } else if (drinkDetected.includes('/')) {
    clarificationNeeded = true;
    reason = 'multiple_drink_candidates';
  }

  const confidence = Math.min(
    0.99,
    Number((drinkConf + (sizeDefaulted ? 0 : 0.07) + (milkDefaulted ? 0 : 0.07)).toFixed(2))
  );

  return {
    name,
    raw_order_text: raw,
    quantity: qty,
    drink,
    size,
    milk,
    defaults_applied: {
      size: sizeDefaulted,
      milk: milkDefaulted,
    },
    confidence,
    clarification_needed: clarificationNeeded,
    clarification_reason: reason,
  };
}

function run() {
  const base = path.resolve(process.cwd(), 'projects', 'coffee-bot');
  const inputCsv = path.join(base, 'samples', 'blank_street_orders_sample.csv');
  const outputDir = path.join(base, 'output');
  const outputJson = path.join(outputDir, 'parsed_orders.json');

  const raw = fs.readFileSync(inputCsv, 'utf8');
  const rows = parseCsv(raw).map(parseRow);

  const summary = {
    total_orders: rows.length,
    clarification_count: rows.filter(r => r.clarification_needed).length,
    auto_default_size_count: rows.filter(r => r.defaults_applied.size).length,
    auto_default_milk_count: rows.filter(r => r.defaults_applied.milk).length,
  };

  fs.mkdirSync(outputDir, { recursive: true });
  fs.writeFileSync(outputJson, JSON.stringify({ summary, orders: rows }, null, 2), 'utf8');

  console.log(`Parsed ${summary.total_orders} rows -> ${outputJson}`);
  console.log(`Clarifications needed: ${summary.clarification_count}`);
}

run();
